kore/src/Kore/Step/Simplification/Overloading.hs

```diff
@@ -27,113 +27,154 @@ import Kore.Internal.Pattern
+{- |
+ If the two constructors form an overload pair, apply the overloading axioms
+ on the terms to make the constructors equal, then retry unification on them.
 
 See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
 
-If the two constructors form an overload pair, then use the sorting information
-for the two overloaded constructors to directly derive the new goals
-(apply the overload axiom right-to-left on the right and retry)
-otherwise, return bottom, as the constructors are incompatible
-
  -}
 overloadedConstructorSortInjectionAndEquals
     :: (SimplifierVariable variable, MonadUnify unifier)
-    => HasCallStack
     => TermSimplifier variable unifier
     -> TermLike variable
     -> TermLike variable
     -> MaybeT unifier (Pattern variable)
-overloadedConstructorSortInjectionAndEquals
-    termMerger
-    first@(App_ firstHead _)
-    (Inj_ secondInj)
-  = do
-    OverloadSimplifier { isOverloaded } <- Simplifier.askOverloadSimplifier
-    Monad.guard (isOverloaded firstHead)
-    overloadedAndEqualsOverloading termMerger first secondInj
-overloadedConstructorSortInjectionAndEquals
-    termMerger
-    (Inj_ firstInj)
-    second@(App_ secondHead _)
+overloadedConstructorSortInjectionAndEquals termMerger firstTerm secondTerm
   = do
-    OverloadSimplifier { isOverloaded } <- Simplifier.askOverloadSimplifier
-    Monad.guard (isOverloaded secondHead)
-    overloadedAndEqualsOverloading termMerger second firstInj
-overloadedConstructorSortInjectionAndEquals
-    termMerger
-    first@(Inj_ inj@Inj { injTo, injChild = App_ firstHead firstChildren})
-    second@(Inj_ Inj { injChild = App_ secondHead secondChildren})
+    eunifier <- Monad.Trans.lift (unifyOverloading (Pair firstTerm secondTerm))
+    case eunifier of
+        Right (Just (Pair firstTerm' secondTerm')) -> Monad.Trans.lift $
+            termMerger firstTerm' secondTerm'
+        Left message -> Monad.Trans.lift $
+            explainAndReturnBottom message firstTerm secondTerm
+        Right Nothing -> empty
+
+{- |
+ Tests whether the pair of terms can be coerced to have the same constructors
+ at the top, and, if so, returns the thus obtained new pair.
+
+ See <https://github.com/kframework/kore/blob/master/docs/2019-08-27-Unification-modulo-overloaded-constructors.md>
+
+ If the overloading rules are not appliable, it would return @Nothing@.
+ However, if it detects an unification failure, it would return a reason
+ for the failure which can be used as an explanation for it.
+
+ Note: this algorithm is used for both matching and unification; that is why
+ the first and second terms in a pair are not interchangeable.
+-}
+unifyOverloading
+    :: MonadSimplify unifier
+    => SimplifierVariable variable
+    => Pair (TermLike variable)
+    -> unifier
+        (Either (Doc ()) (Maybe (Pair (TermLike variable))))
+unifyOverloading termPair = case termPair of
+    Pair
```

Comment:
Could you please add brief comment to each of these cases, explaining what it is checking for?